### PR TITLE
Add composite signal support to PlatformIO.read_signal

### LINKED ISSUE #827: There is no way to perform multiple power measurements.
The singleton geopm::platform_io, once a push_signal is performed, never resets the flag m_is_active to false. A work around is to instantiate a non-public class:

const auto platform_io = std::make_unique<geopm::PlatformIOImp>();
const auto idx = platform_io->push_signal(signal_name, domain_type, domain_index);
[...]

This code is used in a daemon started by systemd, that reads on-demand, i.e. not in a cli tool.

### PR DESCRIPTION
**Pull request by bgeltz**
Fixes #827.
Uses DifferenceCombinedSignal.
Also fixes a few unit tests that were missing config.h.

### CODE REVIEW

**Comment by bgeltz** [comment]
Let's talk a bit on Monday, but I don't see the benefit to this approach.

I think we're limiting the usefulness of DifferenceCombinedSignal by requiring it to have exactly 2 elements.

If we decide to go with this, I need to add a unit test for DifferenceCombined signal.

**Comment by bgeltz** [comment]
Also, I'm not sure why, but the commits in the PR are in a different order than they are in my branch:

```
* d5b74b51 - (HEAD -> bgeltz-issue-827-dcs, bgeltz-public/bgeltz-issue-827-dcs) Use read_signal in geopmread (7 minutes ago) <Brad Geltz>
* 6f0d4cd6 - Add composite signal support to read_signal (7 minutes ago) <Brad Geltz>
* 907d4efb - Add DifferenceCombinedSignal (7 minutes ago) <Brad Geltz>
* 0d74b622 - Spell out PACKAGE for temperature (7 minutes ago) <Brad Geltz>
* e142c621 - (origin/dev, dev) Doxygen documentation and comments nice-to-have (26 hours ago) <Xunjin>
* 3b23bf65 - Removed C++ style comments from C source files (26 hours ago) <Xunjin>
* d63cbeac - Throw error for unknown policy names in json file (3 days ago) <Diana Guttman>
* 0b3ec6dd - Fix a potential bug where policies made of zeros would never be sent by TreeComm (3 days ago) <Diana Guttman>
* 78e6de29 - Add Reduce model region and clean up barrier model region (3 days ago) <Christopher M. Cantalupo>
* c72f5ef8 - Add a barrier model region (3 days ago) <Christopher M. Cantalupo>
```

File: src/CombinedSignal.cpp [hunk 1]
```diff
@@ -109,4 +109,15 @@ namespace geopm
         }
         return result;
     }
+
+    double DifferenceCombinedSignal::sample(const std::vector<double> &values)
+    {
+        if (values.size() != 2) {
```
Comment (bgeltz) [inline src/CombinedSignal.cpp]: Move DEBUG check up above if.

File: src/PlatformIO.cpp [hunk 2]
```diff
@@ -528,8 +519,31 @@ namespace geopm
         double result = NAN;
         auto iogroup = find_signal_iogroup(signal_name);
         if (iogroup == nullptr) {
-            throw Exception("PlatformIOImp::read_signal(): signal name \"" + signal_name + "\" not found",
-                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            std::size_t sep = signal_name.find("_");
+            std::string type = sep != std::string::npos ? signal_name.substr(sep) : "";
+            if (signal_name.find("POWER") != std::string::npos) {
```
Comment (bgeltz) [inline src/PlatformIO.cpp]: Be strict about the index == 0 here if we're sure this starts with "POWER...".

File: src/CombinedSignal.cpp [hunk 3]
```diff
@@ -109,4 +109,15 @@ namespace geopm
         }
         return result;
     }
+
+    double DifferenceCombinedSignal::sample(const std::vector<double> &values)
+    {
+#ifdef GEOPM_DEBUG
+        if (values.size() != 2) {
+            throw Exception("PlatformIOImp::push_signal_temperature(): expected two values to subtract for temperature.",
```
Comment (dianarg) [inline src/CombinedSignal.cpp]: error message is out of date

File: src/CombinedSignal.hpp [hunk 4]
```diff
@@ -73,6 +73,16 @@ namespace geopm
             CircularBuffer<m_sample_s> m_history;
             int m_derivative_num_fit;
     };
+
+    /// @brief Used by PlatformIO for CombinedSignals based on a
+    ///        difference of another signal.
```
Comment (dianarg) [inline src/CombinedSignal.hpp]: difference between two signals

**Review by dianarg — CHANGES_REQUESTED** [review]

**Review by dianarg — DISMISSED** [review]

**Review by dianarg — CHANGES_REQUESTED** [review]
👎 
test if this will be wiped out by force push

**Review by dianarg — APPROVED** [review]
👍

File: src/PlatformIO.cpp [hunk 5]
```diff
@@ -528,8 +519,31 @@ namespace geopm
         double result = NAN;
         auto iogroup = find_signal_iogroup(signal_name);
         if (iogroup == nullptr) {
-            throw Exception("PlatformIOImp::read_signal(): signal name \"" + signal_name + "\" not found",
-                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            std::size_t sep = signal_name.find("_");
+            std::string type = sep != std::string::npos ? signal_name.substr(sep) : "";
+            if (signal_name.find("POWER") == 0) {
```
Comment (cmcantalupo) [inline src/PlatformIO.cpp:524]: Should be

```
if (signal_name.find("POWER") != std::string::npos) {
```

**Pull request merged**